spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraJoinRDD.scala

```diff
@@ -230,18 +235,36 @@ class CassandraJoinRDD[L, R] private[connector](
     countingIterator
   }
 
-  private def fetchIterator(
+  private[rdd] def fetchIterator(
     session: Session,
     bsb: BoundStatementBuilder[L],
-    lastIt: Iterator[L]): Iterator[(L, R)] = {
-
+    leftIterator: Iterator[L]): Iterator[(L, R)] = {
     val columnNamesArray = selectedColumnRefs.map(_.selectedAs).toArray
-    for (leftSide <- lastIt;
-         rightSide <- {
-           val rs = session.execute(bsb.bind(leftSide))
-           val iterator = new PrefetchingResultSetIterator(rs, fetchSize)
-           iterator.map(rowReader.read(_, columnNamesArray))
-         }) yield (leftSide, rightSide)
+    val rateLimiter = new RateLimiter(
+      readConf.throughputJoinQueryPerSec, readConf.throughputJoinQueryPerSec)
+
+    def pairWithRight(left: L): SettableFuture[Iterator[(L, R)]] = {
+      val resultFuture = SettableFuture.create[Iterator[(L, R)]]
+      val leftSide = Iterator.continually(left)
+
+      val queryFuture = session.executeAsync(bsb.bind(left))
+      Futures.addCallback(queryFuture, new FutureCallback[ResultSet] {
+        def onSuccess(rs: ResultSet) {
+          val resultSet = new PrefetchingResultSetIterator(rs, fetchSize)
+          val rightSide = resultSet.map(rowReader.read(_, columnNamesArray))
+          resultFuture.set(leftSide.zip(rightSide))
+        }
+        def onFailure(throwable: Throwable) {
+          resultFuture.setException(throwable)
+        }
+      })
+      resultFuture
+    }
+    val materializedFutures = leftIterator.map(left => {
```

Comment:
Perhaps we want to rename this to something to illustrate that this is the set of query futures and that they may not have their values yet. "Materialized" in my head kept making me think that we expected the results to be here.